api/traces/v1/http.go

```diff
@@ -0,0 +1,293 @@
+package v1
+
+import (
+	"bytes"
+	"compress/flate"
+	"compress/gzip"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"net/http/httputil"
+	"net/url"
+	"strings"
+	"time"
+
+	"github.com/go-chi/chi"
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/prometheus/client_golang/prometheus"
+	"go.opentelemetry.io/contrib/instrumentation/net/http/otelhttp"
+
+	"github.com/observatorium/api/proxy"
+)
+
+const (
+	ReadTimeout = 15 * time.Minute
+)
+
+type handlerConfiguration struct {
+	logger           log.Logger
+	registry         *prometheus.Registry
+	instrument       handlerInstrumenter
+	spanRoutePrefix  string
+	readMiddlewares  []func(http.Handler) http.Handler
+	writeMiddlewares []func(http.Handler) http.Handler
+}
+
+// HandlerOption modifies the handler's configuration.
+type HandlerOption func(h *handlerConfiguration)
+
+// Logger add a custom logger for the handler to use.
+func Logger(logger log.Logger) HandlerOption {
+	return func(h *handlerConfiguration) {
+		h.logger = logger
+	}
+}
+
+// WithRegistry adds a custom Jaeger query for the handler to use.
+func WithRegistry(r *prometheus.Registry) HandlerOption {
+	return func(h *handlerConfiguration) {
+		h.registry = r
+	}
+}
+
+// WithHandlerInstrumenter adds a custom HTTP handler instrument middleware for the handler to use.
+func WithHandlerInstrumenter(instrumenter handlerInstrumenter) HandlerOption {
+	return func(h *handlerConfiguration) {
+		h.instrument = instrumenter
+	}
+}
+
+// WithSpanRoutePrefix adds a prefix before the value of route tag in tracing spans.
+func WithSpanRoutePrefix(spanRoutePrefix string) HandlerOption {
+	return func(h *handlerConfiguration) {
+		h.spanRoutePrefix = spanRoutePrefix
+	}
+}
+
+// WithReadMiddleware adds a middleware for all read operations.
+func WithReadMiddleware(m func(http.Handler) http.Handler) HandlerOption {
+	return func(h *handlerConfiguration) {
+		h.readMiddlewares = append(h.readMiddlewares, m)
+	}
+}
+
+// WithWriteMiddleware adds a middleware for all write operations.
+func WithWriteMiddleware(m func(http.Handler) http.Handler) HandlerOption {
+	return func(h *handlerConfiguration) {
+		h.writeMiddlewares = append(h.writeMiddlewares, m)
+	}
+}
+
+type handlerInstrumenter interface {
+	NewHandler(labels prometheus.Labels, handler http.Handler) http.HandlerFunc
+}
+
+type nopInstrumentHandler struct{}
+
+func (n nopInstrumentHandler) NewHandler(labels prometheus.Labels, handler http.Handler) http.HandlerFunc {
+	return handler.ServeHTTP
+}
+
+// NewV2APIHandler creates a trace query handler for Jaeger V2 HTTP queries
+func NewV2APIHandler(read *url.URL, opts ...HandlerOption) http.Handler {
+	if read == nil {
+		panic("missing Jaeger read url")
```

Comment:
We should never really hit this, since we're checking if `read` is nil in the previous method in the chain, isn't it? I'm wondering whether we need the check here.

Reply:
The check is here because due to another bug (which was never committed) the caller passed `nil`.  This function passed that nil to `proxy.MiddlewareSetUpstream()`, which created a closure containing the nil.  The closure was later invoked in the scope of an HTTP interceptor, leading to a stack trace with no hint where the nil came from.

It has been my experience it is far better to catch invalid parameters near the call stack where they are made.

Your comment is correct.  Do you want me to remove the check?

Reply:
Not a must from my side, I'm fine with it :+1: